src/main/java/com/microsoft/azure/proton/transport/proxy/impl/ProxyChallengeProcessorImpl.java

```diff
@@ -0,0 +1,170 @@
+package com.microsoft.azure.proton.transport.proxy.impl;
+
+import com.microsoft.azure.proton.transport.proxy.ProxyChallengeProcessor;
+
+import javax.xml.bind.DatatypeConverter;
+import java.io.UnsupportedEncodingException;
+import java.net.*;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class ProxyChallengeProcessorImpl implements ProxyChallengeProcessor {
```

Comment:
There should be multiple implementations of the ProxyChallengeProcessor interface, one for each auth scheme we support: DigestProxyChallengeProcessorImpl, BasicProxyChallengeProcessorImpl. process() should look at the challenge, select the preferred scheme, and instantiate the class which supports that scheme. This design allows us to cleanly add new schemes later.

Reply:
Thank you, James. The PR is updated with the new suggest changes.

Reply:
Any updates of reviewing, we are trying to get this PR approved and merged ASAP. Please let me know if there is anything I can do to make it happen. Free feel to ping me.
@JamesBirdsall @SreeramGarlapati

Reply:
It looks good to me. I'll merge it tomorrow morning unless Sreeram has something more to say. I think we're at the point where all the major pieces are in place and any tweaks can be handled as separate PRs. Thanks Shawn!

Reply:
Thank you, James.